### PR TITLE
PWN-4629 - Main screen polishing

### PR DESCRIPTION
**Pull request by 1sakov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-4629

## Description of Work

Added diff utils for main screen
Now we don't hide hidden tokens, we are not showing them at all

## Screenshots (Optional)

Screenshot #1 | Screenshot #2 | Screenshot #3
:------------:|:-------------:|:------------:
[Paste Pic 1] | [Paste Pic 2] | [Paste Pic 3]

### CODE REVIEW

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt [hunk 1]
```diff
@@ -44,32 +54,46 @@ class TokenAdapter(
...
-        isHidden: Boolean,
-        list: List<HomeElementItem>,
-        state: VisibilityState
-    ) = list.toMutableList().apply {
-        if (isHidden) {
-            add(0, HomeElementItem.Action(state))
-        } else {
-            add(0, HomeElementItem.Title(R.string.home_tokens))
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int, payloads: MutableList<Any>) {
+        val fields = (payloads.firstOrNull() as? Set<*>)?.filterIsInstance<String>()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt:66]: god damn
that's why we need delegates  😃

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt [hunk 2]
```diff
@@ -44,32 +54,46 @@ class TokenAdapter(
...
+        val visibleTokens = groups[false].orEmpty()
+
+        result += visibleTokens
+        result += Action(state)
+        result += hiddenTokens
+
+        return result
+    }
+
     private fun getDiffCallback(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt:97]: let's extract this diff to a class? it's getting huge

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt (hunk 1 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt]: I made a minor change. 
It should look better

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt (hunk 2 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt]: agree, nice point

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomeFragment.kt [hunk 3]
```diff
@@ -299,4 +299,8 @@ class HomeFragment :
         presenter.clearTokensCache()
         super.onDestroy()
     }
+
+    fun updateTokensIfNeeded() {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomeFragment.kt:303]: private?

**Review by gslevinkov — APPROVED** [review]

**Pull request merged**